### PR TITLE
Implemented MultiChannel logic, calculate_channel_urls and get_channe…

### PR DESCRIPTION
**Pull request by JohanMabille**
…l_urls

I started reimplementing all the MultiChannel logic with composite pattern and so on until I realize we just don't need them.

### CODE REVIEW

**Comment by JohanMabille** [comment]
What's going on here? The build is systematically cancelled.

**Comment by wolfv** [comment]
I haven't cancelled anything. Either github is has issues or they want us to pay for CI time?

File: test/test_channel.cpp [hunk 1]
```diff
@@ -72,5 +79,27 @@ namespace mamba
         EXPECT_EQ(urls2[0], "https://conda.anaconda.org/conda-forge/win-64");
         EXPECT_EQ(urls2[1], "https://conda.anaconda.org/conda-forge/noarch");
     }
+
+    TEST(Channel, calculate_channel_urls)
+    {
+        std::vector<std::string> urls = { "conda-forge", "defaults" };
+        std::vector<std::string> res = calculate_channel_urls(urls, true);
+        EXPECT_EQ(res.size(), 6);
+        EXPECT_EQ(res[0], "https://conda.anaconda.org/conda-forge/linux-64");
```
Comment (wolfv) [inline test/test_channel.cpp:88]: yeah! :tada:

File: src/channel.cpp [hunk 2]
```diff
@@ -518,11 +518,85 @@ namespace mamba
...
+            names.reserve(channel_names.size() + 1 + ctx_channels.size());
+            if (use_local)
+            {
+                names.push_back(LOCAL_CHANNELS_NAME);
+            }
+            std::copy(channel_names.begin(), channel_names.end(), std::back_inserter(names));
+            if (prepend)
+            {
+                std::copy(ctx_channels.begin(), ctx_channels.end(), std::back_inserter(names));
+            }
```
Comment (wolfv) [inline src/channel.cpp:591]: what happens if prepend is false?

Comment (wolfv) [reply src/channel.cpp:591]: ah so the opposite of prepend is not append, it's just to ignore the ctx_channels?

Comment (JohanMabille) [reply src/channel.cpp:591]: Nothing.

Comment (JohanMabille) [reply src/channel.cpp:591]: Btw I kept the name from conda, but I think we should rename it into something more explicit, like "append_context_channel".

Also in our case, the channels_names and the ctx_channels are exactly the same (the context always contains those from the command line + configuration), and I wonder if we'll have a use case where we pass different channel names than those from the command line.

**Pull request merged**